src/plugins/proctracer/proctracer.cpp

```diff
@@ -0,0 +1,390 @@
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <glib.h>
+#include <libvmi/libvmi.h>
+#include <json-c/json.h>
+#include "private.h"
+
+static event_response_t trace_cb(drakvuf_t drakvuf, drakvuf_trap_info_t *info){
+    trace_info *ti=(trace_info*)info->trap->data;
+    proctracer *p=(proctracer*)ti->p;
+    if (p->trace_status.find(info->regs->cr3) == p->trace_status.end())
+        return 0;
+    switch(p->format){
+        case OUTPUT_CSV:
+            printf("proctracer,trace,0x%lx,\"%s\",0x%lx\n",info->regs->cr3, ti->mod_name, ti->offset);
+            break;
+        default:
+        case OUTPUT_DEFAULT:
+            printf("[PROCTRACER] Trace point hit - CR3: 0x%lx Module: %s Offset: 0x%lx \n",info->regs->cr3, ti->mod_name, ti->offset);
+            break;
+    }
+    
+    p->trace_statistics[info->trap_pa]+=1;
+    if (p->ccov_limit > 0 && p->trace_statistics[info->trap_pa] > p->ccov_limit){
+        drakvuf_remove_trap(drakvuf,info->trap,NULL);
+    }
+    return 0;
+}
+
+static event_response_t exit_cb(drakvuf_t drakvuf, drakvuf_trap_info_t *info) {
+    proctracer *p = (proctracer*)info->trap->data;
+    if (p->trace_status.find(info->regs->cr3) == p->trace_status.end())
+        return 0;
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+
+    if (p->trace_status.find(info->regs->cr3) == p->trace_status.end()){
+        drakvuf_release_vmi(drakvuf);
+        return 0;
+    }
+
+    list<mod_info*> module_traps=p->trace_status[info->regs->cr3];
+    for (mod_info* mi: module_traps){
+        for (drakvuf_trap_t* dt: mi->traps){
+            trace_info *ti=(trace_info*)dt->data;
+            free(ti->mod_name);
+            drakvuf_remove_trap(drakvuf,dt,NULL);
+        }
+        delete mi;
+    }
+    p->trace_status.erase(info->regs->cr3);
+
+    drakvuf_release_vmi(drakvuf);
+    return 0;
+}
+
+static bool add_trace_points(proctracer* p, drakvuf_t drakvuf, vmi_instance_t vmi, addr_t process, addr_t cr3, char *proc_name){
+    // [TODO] This is basically duplicate code from inject_traps_modules() - Maybe the library func should provide a callback parameter?
+    // [TODO] Refactored from cb(), can we have a cleaner function signature?
+
+    addr_t module_list=0;
+    bool traps_ok = true;
+
+    vmi_pid_t pid;
```

Comment:
Using libvmi's _va function's is generally unsafe as it relies on Windows' process-list. It is also slower then using access_context_t with the dtb provided, so it should be used whenever possible.